Kubernetes metadata retrieval retry logic

carkmorwin opened the pull request:
This is a messy first pass but wanted to get some feedback on implementation logic here. 

The strategy here is to first try and get all of the metadata we can, and ship that. If we are able to communicate with the API, then great. We then continue with the happy path (apply filter, if no match, add metadata, follow file and ship log lines).

However, if we cannot communicate with the API at that time, we start a goroutine to keep trying. For now, this will try forever (and ultimately exponential backoff). If/when we are able to communicate with the API, we should then act as if we are on the happy path (apply filter, if no match, add metadata, follow file and ship log lines). The catch here is that we need a way to notify the already started Forwarder that you can stop. In the first case, we would never create one, but in our background retry case, we need a channel. 

I'm not crazy about passing around another channel, but we do need a way to tell a particular file forwarder they can stop. And this differs from the quit channel because one is required per forwarder, vs a global one to indicate we are shutting down the entire agent.

Any help/thoughts/feedback here would be much appreciated, and I'll be cleaning up the current code in the near future.

DavidAntaramian commented:
Once you merge #134 can you rebase this? I'll review it then so I make sure I'm reviewing the correct code

carkmorwin commented:
@DavidAntaramian done. This PR now only includes relevant changes.

DavidAntaramian commented on main.go:
```diff
@@ -361,30 +364,94 @@ func runCaptureKube(ctx *cli.Context) error {
+						// https://blog.golang.org/go-concurrency-patterns-timing-out-and
+						stop = make(chan bool, 1)
+
+						go func() {
```
The nesting here is getting far too deep. Can you extract this logic out into a function that can be called in this context?

carkmorwin replied:
👍

DavidAntaramian commented on main.go:
```diff
@@ -361,30 +364,94 @@ func runCaptureKube(ctx *cli.Context) error {
+							// https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+							attempts := 1
+							base := 1
+							rand.Seed(time.Now().UTC().UnixNano())
```
First, why is this necessary?

Second, does this function affect the entire runtime or just this go routine?

carkmorwin replied:
So the goal here was to add a random factor to our wait time to avoid all requests querying the api at the same time. It might be a bit of a premature optimization but lets say if we are watching 1k files, then 1k requests will all attempt to be made at the same time against the api. With the random factor there is some skew, and a better chance requests succeed asap. 

Regardless, that rand call is incorrect and will be fixed. We should be getting a separate rand object per goroutine.

DavidAntaramian replied:
That answers my questions. As you said, being able to replicate the test seed is beneficial, so I'd like to make sure we keep that possibility.

DavidAntaramian commented on main.go:
```diff
@@ -361,30 +364,94 @@ func runCaptureKube(ctx *cli.Context) error {
+											fileConfig.Path, filter)
+
+										// Send quit signal to forwarder of this file
+										stop <- true
```
`close(stop)` sends `0` anyways, so is this necessary if you're not explicitly acting on `true`?

carkmorwin replied:
Nope. Wasn't sure how close behaved but if it sends 0 then that's perfect.

carkmorwin commented:
@DavidAntaramian Sorry for the larger than expected set of changes/refactor, but I think readability is improved here, especially from `main.go`.

carkmorwin commented:
@DavidAntaramian Not sure. The fact that we proxy a connection to the api makes me want to keep trying vs if we were hitting the api directly then we would bail. The reason here is that if you can't talk to the API then something is wrong with the node, as pods wouldnt get scheduled etc. 

Additionally, let's say I cut the proxy and put it back. I'd like to always keep trying in that case, or basically once we get the proxy back we collect our data. That time window could be anything, and it's outside of our control. I'd like the agent to be resilient without having to restart it, but I could go either way here.

Also maybe a different approach in terms of retries makes more sense as we only need to succeed once per pod boot, versus regularly for all files. 

Maybe we retry exponentially and then cap at and keep trying every 10 minutes. Let's talk offline about this, would prefer to hash out a solid approach.

carkmorwin commented:
Per offline discussion, going to add an upper bound to retry time and then dev is done.

carkmorwin commented:
@DavidAntaramian Just pushed last commit for work being done on this PR. If you can give it just one final review, that would be appreciated.

DavidAntaramian commented on kubernetes_metadata.go:
```diff
@@ -234,6 +234,10 @@ func (client *KubernetesClient) Available() bool {
 //GetMetadataFromAPIWithBackoff Queries the Kubernetes API for metadata and if available updates the *KubernetesContext.
 // If the API is unavailable, retries are performed with exponential backoff.
 func (client *KubernetesClient) GetMetadataFromAPIWithBackoff(context *KubernetesContext, rand *rand.Rand) {
+	var sleepLimit int // Upper limit on sleepDuration in seconds
```
Can you explicitly define this as

```go
sleepLimit := 0
```

carkmorwin commented:
@DavidAntaramian As far a changes to the sleeping code, I removed the base argument to since it was always set to 1, and added comments explaining choices. I think things are explicit enough to be understood by someone looking at code, but let me know what you think.

DavidAntaramian commented:
@carkmorwin That's way better. The comments help me follow the logic 👍 really happy about this

carkmorwin commented:
@DavidAntaramian 🎉 Going to squash and merge this.
